### ISSUE TITLE
enh: use kqueue on BSD

### ISSUE THREAD

**Issue by asomers** [alexcrichton/tokio-signal#20]
Instead of the thread+handler+pipe approach, tokio-signal could directly use a kqueue filter on all of the BSDs.  Basically, you just block that signal from your process, then register the `EVFILT_SIGNAL` kqueue filter, and every signal delivery will show up directly in your kqueue.  No threads, handlers, or pipes required.  It's supported by all the major BSDs, plus OSX.

https://www.freebsd.org/cgi/man.cgi?query=kqueue&sektion=2
http://netbsd.gw.com/cgi-bin/man-cgi?kqueue++NetBSD-current
https://man.openbsd.org/kqueue.2
https://leaf.dragonflybsd.org/cgi/web-man?command=kqueue&section=2
https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man2/kqueue.2.html

**Comment by vorner** [comment]
I don't know BSDs, but is there a reliable way to block the signal on the whole *process*, including *all* the threads? Otherwise you need to make sure it is blocked *before* any threads start and it's hard to guarantee when you're a library, not the program.

**Comment by ipetkov** [comment]
@vorner  According to the man pages it seems like the event is delivered if the signal is sent to the process, regardless if a particular thread received/handled it or if has been ignored. Seems like it may be possible!

> EVFILT_SIGNAL    The filter will record all attempts to deliver a signal to a process, even if the signal has been marked as SIG_IGN. Event notification happens after normal signal delivery processing. data returns the number of times the signal has occurred since the last call to kevent(). This filter automatically sets the EV_CLEAR flag internally.

Though I think we should be careful about this caveat in the FreeBSD version since it may have implications for `tokio-process`:

> The filter will record all attempts to deliver a signal to a process, even if the signal has been marked as SIG_IGN, except for the SIGCHLD signal, which, if ignored, won't be recorded by the filter.

**Comment by asomers** [comment]
I agree with @ipetkov that tokio-signal should work correctly regardless of whether the signal is blocked by all threads.  The `SIGCHLD` exception is weird, but it shouldn't matter for `tokio-process`, because that crate ought to be using `EVFILT_PROC` instead of signals.

**Comment by ipetkov** [comment]
@asomers great point, I wasn't aware of `EVFILT_PROC`! Seems like `tokio-process` could depend on `mio` directly for kqueue systems, but I think we'll need changes in `mio` to enable this
